src/scipp/signal/__init__.py

```diff
@@ -0,0 +1,105 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock
+"""Sub-package for signal processing such as smoothing.
+
+This subpackage provides wrappers for a subset of functions from
+:py:mod:`scipy.signal`.
+"""
+from dataclasses import dataclass
+from numpy import ndarray
+
+from ..core import array, identical, islinspace, DataArray, Variable
+from ..core import UnitError, CoordError
+from ..units import one
+from ..compat.wrapping import wrap1d
+
+
+@dataclass
+class SOS:
+    """
+    Second-order sections representation returned by :py:func:`scipp.signal.butter`.
+
+    This is used as input to :py:func:`scipp.signal.sosfiltfilt`.
+    """
+    coord: Variable
+    sos: ndarray
+
+
+def _frequency(coord: Variable) -> Variable:
+    if not islinspace(coord).value:
+        raise CoordError("Data is not regularly sampled, cannot compute frequency.")
+    return (len(coord) - 1) / (coord[-1] - coord[0])
+
+
+def butter(da: DataArray, dim: str, *, N, Wn, **kwargs):
+    """
+    Butterworth digital and analog filter design.
+
+    Design an Nth-order digital or analog Butterworth filter and return the filter
+    coefficients.
+
+    This is intended for use with :py:func:`scipp.signal.sosfiltfilt`. See there for
+    an example.
+
+    This is a wrapper around :py:func:`scipy.signal.butter`. See there for a
+    complete description of parameters. The differences are:
+
+    - Instead of a sampling frequency fs, this wrapper takes a data array ``da`` and a
+      dimension label ``dim`` as input. The sampling frequency is then computed from
+      ``da.coords[dim]``. Only inputs with regular sampling are supported.
+    - The critical frequency or frequencies must be provided as a variable with correct
+      unit.
+    - Only 'sos' output is supported.
+
+    :seealso: :py:func:`scipp.signal.sosfiltfilt`
+    """
+    coord = da.coords[dim]
+    fs = _frequency(coord).value
+    if not Wn.unit == one / coord.unit:
+        raise UnitError(
+            f"Critical frequency unit '{Wn.unit}' does not match sampling unit "
+            f"'{one / coord.unit}'")
+    import scipy.signal
+    return SOS(coord=coord.copy(),
+               sos=scipy.signal.butter(N=N, Wn=Wn.values, fs=fs, output='sos',
+                                       **kwargs))
+
+
+@wrap1d(keep_coords=True)
+def sosfiltfilt(da: DataArray, dim: str, *, sos: SOS, **kwargs) -> DataArray:
```

Comment:
Even though this would break with mirroring the scipy interface, should this function be a method of `SOS`? This way, if we introduce other filter encodings, it would not be possible to use the wrong `filter` function. (Not possible right now either because `sos.sos` would fail but that is harder to understand for a user.)

Reply:
Hmm, interesting idea. So you would have something like:

```python
da = butter(da, 'dim', N=4, Wn=cutoff).filtfilt()
```
?

Reply:
Above assumes that `da` is used for both, setting up the filter, and as filteree... in general that may not be the case?

Reply:
> in general that may not be the case?

Hence I would have said
```.py
da = butter(da, 'dim', N=4, Wn=cutoff).filtfilt(da, 'dim')
```
Looks odd that you have to repeat `da` and `dim` but this is anyway the case.

We could of course make the args of `filtilf` default to the ones given to `butter`.

Reply:
Coming back to this discussion: I am now quite unhappy with the interface. Why should `butter` take a data array as an argument and not just a coord.

Current solutions has a couple of problems and limitations:

- `da` is not needed by `butter`, this is odd and confusing.
- Forces coord to be stored in a data array (potentially dummy) to be able to call `butter`.
- Forces coord to be stored in data array to be filtered.

I see the following solutions, but neither of them is clear right now:

1. `butter` should only take a coord as input.
2. Unified function that does both the filter setup (`butter`) and applies `sosfiltfilt`.
3. As 2.) but similar to discussion above, returning `SOS` with a method?

Reply:
How about:

```python
var = butter(coord, N=4, Wn=cutoff).filtfilt(var, 'dim')  # should be possible to operate on coord
da = butter(coord, N=4, Wn=cutoff).filtfilt(da, 'dim')  # verify coord (if present)
da.data = butter(coord, N=4, Wn=cutoff).filtfilt(da.data, 'dim')  # bypass coord verification
```

... but maybe when passed a variable, this should return a data array, with the coord given to `butter`?

Reply:
Isn't the point of DataArrays to keep coords and data together. This seems a bit like the situation in C with separate array and length function arguments. It is best to keep them together to make sure they actually correspond to each other. Here, it is not quite as important because there is no risk of segfaulting. But I still think that data and coords should not be separated, at least in the public API.

I do however agree that your suggestion is more flexible and it can be annoying to be forced to create temporary DataArrays.
Also, the current API only works with dim-coords. But there is in principle no reason why it wouldn't work with other coords as well.

Barring my first paragraph (as far as inputs are concerned), your last suggestion looks good. But I would still return a DataArray even when the input is a Variable (because of the first paragraph).

Reply:
So in summary: `butter` should take a coord, use it for verification (if applicable), and filtering should always return a data array.